src/zarr/core/metadata/v3.py

```diff
@@ -399,7 +482,24 @@ def from_dtype(cls, dtype: np.dtype[Any]) -> DataType:
             "<u2": "uint16",
             "<u4": "uint32",
             "<u8": "uint64",
+            "<f2": "float16",
             "<f4": "float32",
             "<f8": "float64",
+            "<c8": "complex64",
+            "<c16": "complex128",
         }
         return DataType[dtype_to_data_type[dtype.str]]
+
+
+def parse_dtype(data: npt.DTypeLike) -> np.dtype[Any]:
```

Comment:
do we need a narrow annotation for the input? the idea for `parse_dtype` is to accept _anything_ and error if it's invalid. if we use a narrow annotation for the input (e.g., `npt.DTypeLike`) then, for inputs with unknown types (like stuff we read from JSON), then we need some type checking / casting in advance to appease mypy

Reply:
Reading the docs, this basically is an object type:

```
A Union representing objects that can be coerced into a dtype.

Among others this includes the likes of:

type objects.
Character codes or the names of type objects.
Objects with the .dtype attribute.

```

Reply:
I just checked the output of `json.loads` and it's typed as `Any`, so I guess we won't need any type narrowing for inputs to this function then, but if we start needing that then we should consider widening the annotation